### PR TITLE
Add spring-data sample with Kotlin

### PR DESCRIPTION
**Pull request by christophstrobl**
Add a copy of the `data-jpa` sample using _Kotlin_ to verify the specific reflection needs when using _Kotlin_ (see: spring-projects/spring-data-commons#2717)

Note: This sample will fail till the issue in `data-commons` is resolved.

### CODE REVIEW

File: data-jpa-kotlin/src/main/kotlin/com/example/data/jpa/kotlin/AutorRepository.kt [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package com.example.data.jpa.kotlin
+
+import com.example.data.jpa.kotlin.model.Author
+import org.springframework.data.jpa.repository.JpaRepository
+import org.springframework.data.jpa.repository.Query
+import java.util.*
+
+interface AuthorRepository : JpaRepository<Author?, Long?> {
+
+	fun findByNameContainingIgnoreCase(partialName: String?): Optional<Author?>?
```
Comment (mhalbritter) [inline data-jpa-kotlin/src/main/kotlin/com/example/data/jpa/kotlin/AutorRepository.kt:10]: A nullable `Optional` doesn't seem right. Maybe just `Author?`?

File: data-jpa-kotlin/src/main/kotlin/com/example/data/jpa/kotlin/AutorRepository.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package com.example.data.jpa.kotlin
+
+import com.example.data.jpa.kotlin.model.Author
+import org.springframework.data.jpa.repository.JpaRepository
+import org.springframework.data.jpa.repository.Query
+import java.util.*
+
+interface AuthorRepository : JpaRepository<Author?, Long?> {
```
Comment (mhalbritter) [inline data-jpa-kotlin/src/main/kotlin/com/example/data/jpa/kotlin/AutorRepository.kt:8]: `<Author?, Long?>`: can that be `<Author, Long>`?

File: data-jpa-kotlin/src/main/kotlin/com/example/data/jpa/kotlin/AutorRepository.kt [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+	fun findByNameContainingIgnoreCase(partialName: String?): Optional<Author?>?
+
+	@Query("SELECT a FROM Author a WHERE a.name = :name")
+	fun queryFindByName(name: String?): Optional<Author?>?
```
Comment (mhalbritter) [inline data-jpa-kotlin/src/main/kotlin/com/example/data/jpa/kotlin/AutorRepository.kt:13]: I think the `name` shouldn't be nullable here.

File: data-jpa-kotlin/src/main/kotlin/com/example/data/jpa/kotlin/model/Autor.kt [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+@Entity
+class Author {
+
+	@Id
+	@GeneratedValue
+	var id: Long? = null
+	var name: String? = null
+
+	@OneToMany(cascade = [CascadeType.ALL])
+	var books: Set<Book>? = null
```
Comment (mhalbritter) [inline data-jpa-kotlin/src/main/kotlin/com/example/data/jpa/kotlin/model/Autor.kt:15]: Make it non-nullable and assign `emptySet` to it? Does that work with JPA?

**Review by mhalbritter — CHANGES_REQUESTED** [review]

**Review by sdeleuze — CHANGES_REQUESTED** [review]
Please add a JPA callback like `@PreRemove` to test spring-projects/spring-framework#29348 with Kotlin.

**Comment by christophstrobl** [comment]
@mhalbritter applied some changes

**Comment by mhalbritter** [comment]
Thanks, going to polish it.

**Pull request merged**